addon/initializers/m3-store.js

```diff
@@ -55,17 +56,15 @@ export function extendStore(Store) {
     _preloadSingleResource(data) {
       let modelName = dasherize(data.type);
       let internalModel = this._internalModelForId(modelName, data.id);
-      let preloadData = {};
+      let isUpdate = internalModel.currentState.isEmpty === false;
 
-      // currently internal model expects a flattened preload structure instead of json-api
-      if (typeof data.attributes === 'object' && data.attributes !== null) {
-        Object.assign(preloadData, data.attributes);
-      }
-      if (typeof data.relationships === 'object' && data.relationships !== null) {
-        Object.assign(preloadData, data.relationships);
-      }
+      internalModel.setupData(data);
 
-      internalModel.preloadData(preloadData);
+      if (isUpdate === true) {
+        this.recordArrayManager.recordDidChange(internalModel);
+      } else {
+        internalModel.currentState = EmptyState;
```

Comment:
Do you need to explicitly set the state here? I assumed it is correctly preinitialized.

Reply:
when actually pre-loading (e.g. the non-update case), `setupData` will move the model into a `loaded` state, so we must reset

Reply:
Ah, yeah, forgot about this required hack :)